### PR TITLE
Migration to Jakarta EE 10

### PR DESCRIPTION
**Pull request by essobedo**
## Motivation

Many frameworks like Spring Boot, Quarkus, Camel... have made the Jakarta EE 10 migration which makes the lib incompatible with them. The goal of these changes is to migrate the lib to make it compatible again with the frameworks.

## Modifications:

* Replace the Javax dependencies with the corresponding Jakarta dependencies
* Replace the `javax` imports with the corresponding `jakarta` imports
* Reorganize the imports to have them at the right place

### CODE REVIEW

**Comment by essobedo** [comment]
@zubri Hi, what do you think of this improvement? Thx in advance for any feedbacks 🙏

**Comment by zubri** [comment]
Task linked: [CU-2np2uh3 GH-56 ⁃ JAXB Upgrade Plan](https://app.clickup.com/t/2np2uh3)

File: .settings/org.eclipse.jdt.ui.prefs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 eclipse.preferences.version=1
 org.eclipse.jdt.ui.ignorelowercasenames=true
-org.eclipse.jdt.ui.importorder=\#;;java;javax;
+org.eclipse.jdt.ui.importorder=\#;;java;jakarta;
```
Comment (sman-81) [inline .settings/org.eclipse.jdt.ui.prefs:3]: I'd prefer to only give special treatment in import order to java* packages.

File: src/main/java/com/prowidesoftware/swift/constraints/BicConstraint.java [hunk 2]
```diff
@@ -22,8 +22,8 @@
 import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 
-import javax.validation.Constraint;
-import javax.validation.Payload;
+import jakarta.validation.Constraint;
```
Comment (sman-81) [inline src/main/java/com/prowidesoftware/swift/constraints/BicConstraint.java:25]: See comment about import order above.

File: src/main/java/com/prowidesoftware/swift/constraints/BicValidator.java [hunk 3]
```diff
@@ -19,8 +19,8 @@
 import com.prowidesoftware.swift.model.BicValidationResult;
 import org.apache.commons.lang3.StringUtils;
 
-import javax.validation.ConstraintValidator;
-import javax.validation.ConstraintValidatorContext;
+import jakarta.validation.ConstraintValidator;
```
Comment (sman-81) [inline src/main/java/com/prowidesoftware/swift/constraints/BicValidator.java:22]: See comment about import order above. `jakarta` should sort between `com` and `org` in this example.

File: src/main/java/com/prowidesoftware/swift/model/AbstractSwiftMessage.java [hunk 4]
```diff
@@ -23,7 +23,6 @@
 import com.prowidesoftware.swift.utils.Lib;
 import org.apache.commons.lang3.ArrayUtils;
 import org.apache.commons.lang3.StringUtils;
-import org.apache.commons.lang3.Validate;
```
Comment (sman-81) [inline src/main/java/com/prowidesoftware/swift/model/AbstractSwiftMessage.java:26]: Change seems unrelated.

File: src/test/java/com/prowidesoftware/swift/io/parser/SwiftParserConsumeBlockLenientTest.java [hunk 5]
```diff
@@ -319,7 +319,7 @@ public void testSimpleBlockConsumerBlock3_3_KnownToFail() throws IOException {
         assertEquals("00112233", b.getTagValue("108"));
         assertTrue(b.containsTag("4"));
         assertEquals("foobar", b.getTagValue("4"));
-        assertEquals(b.getUnparsedTextsSize(), new Integer(1));
```
Comment (sman-81) [inline src/test/java/com/prowidesoftware/swift/io/parser/SwiftParserConsumeBlockLenientTest.java:322]: Seems unrelated.

File: build.gradle [hunk 6]
```diff
@@ -36,7 +36,7 @@ project.version = "${SRU}-${scmVersion.version}"
 
 java {
     toolchain {
-        languageVersion.set(JavaLanguageVersion.of(8))
```
Comment (sman-81) [inline build.gradle:39]: Could move to Java 11, then release, then migrate to Jakarta, release again, to decouple these two breaking changes.

**Review by sman-81 — COMMENTED** [review]
I am not too familiar with the `axion-release-plugin`.
How is scmVersion.version set and where can a version suffix such as `-rc1` be added

File: build.gradle (hunk 6 above)
Comment (essobedo) [reply build.gradle:39]: Actually, they are coupled since Jakarta EE 10 classes are compiled with Java 11 so Java 11 is the minimal version

File: src/test/java/com/prowidesoftware/swift/io/parser/SwiftParserConsumeBlockLenientTest.java (hunk 5 above)
Comment (essobedo) [reply src/test/java/com/prowidesoftware/swift/io/parser/SwiftParserConsumeBlockLenientTest.java:322]: it is related to the migration to Java 11

File: build.gradle (hunk 6 above)
Comment (essobedo) [reply build.gradle:39]: Sorry, I misread the comment, yes what you propose would work

File: src/test/java/com/prowidesoftware/swift/io/parser/SwiftParserConsumeBlockLenientTest.java (hunk 5 above)
Comment (sman-81) [reply src/test/java/com/prowidesoftware/swift/io/parser/SwiftParserConsumeBlockLenientTest.java:322]: I see, to avoid the deprecation warning.
This change does not directly have to do with the migration Jakarta. I am much in favour of moving to Java 11 first, release, then migrate to Jakarta.

File: build.gradle (hunk 6 above)
Comment (sman-81) [reply build.gradle:39]: There is value in splitting the tasks Java 11 and Jakarta I think. I would propose to:

- Upgrade the project to Java 11 (including code style/deprecation fixes)
- Release (major version): First Java 11 release still using javax
- Upgrade to Jakarta (your PR)
- Release (major version): Java 11 release, now using Jakarta

This approach would offer enhanced flexibility to the user base.
There may be users welcoming Java 11 but wanting to avoid Jakarta for the time being.

Even Java 11 by itself feels ancient (active support ends this year), we must keep in mind usage is in financial industry mostly which (from my experience) is not an early adopter.
Upgrading to Java 11 should be a well-publicized step. The same goes for the switch to Jakarta. Both are breaking changes.

> So you prefer 2 PRs instead of one? Or 2 different commits for the same PR could be good enough?
Different PRs would be my preference. I don't do releases on this project, so I would leave that up to Sebastian (@zubri).

cheers

File: src/main/java/com/prowidesoftware/swift/constraints/BicConstraint.java (hunk 2 above)
Comment (zubri) [reply src/main/java/com/prowidesoftware/swift/constraints/BicConstraint.java:25]: We normally do organize import with the IDE formatter and leave the order that is produced by it.

File: build.gradle (hunk 6 above)
Comment (zubri) [reply build.gradle:39]: I understand the concern of the finance market adoption that is normally slow. However, splitting into two major releases for Java 11 and for Jakarta migration might be too much to handle. This, considering that besides the technology upgrades we have the yearly standards update.

Current standard is SRU2022 and we have a commitment to release the new standards SRU2023 by May. The SRU2023 standards go live at SWIFT by November 2023.

So between May and December we already have to maintain two active branches for current SRU2022 and upcoming SRU2023. If on top we added two extra branches for Java 11 and Jakarta, that becomes, 4 active branches to maintain until November 2023.

All in all, my initial idea would be to target these releases:

- SRU2022-9.3.* -> current standard version that we have to maintain until November 2023.
- SRU2022-10.0.0 -> including both the upgrade to Java 11 and the migration to Jakarta. With alpha-1, alpha-2, etc suffixes until it is stable and tested. For early adopters of the migrated packages, for current SRU until November 2023.
- SRU2023-10.1.0 -> By May. For early testers of the new standards along the migration to Java 11 and Jakarta.

This scheme would give current users enough time to update (until November 2023). And for early adopters, they will be able to do it for both the current and the upcoming standard messages.

Also in the above scheme, major version clearly reflects the Java 11 and Jakarta breaking change. While the minor version, as usual, reflects the yearly standards update.

Finally, all this applies for both projects, the Prowide Core and the Prowide ISO 20022 (the one with the higher impact)

**Comment by zubri** [comment]
> I am not too familiar with the axion-release-plugin.
> How is `scmVersion.version` in `build.gradle` set and where can a version suffix such as `-rc1` be added

The idea is to use alpha-1, alpha-2, etc... for this. The axion plugin uses automatic versioning based on the SCM tags. However, until this is finally release we can just override the version in the build.gradle with a hardcoded value. Then, when release, it is changed back to the automatic versioning.

File: build.gradle (hunk 6 above)
Comment (essobedo) [reply build.gradle:39]: Moreover from a pure end-user perspective, proposing a version compiled with Java 11 has minimal benefit compared to what is already possible with this lib that is already compatible with a JRE 11 and even a JRE 17.

File: src/main/java/com/prowidesoftware/swift/constraints/BicConstraint.java (hunk 2 above)
Comment (sman-81) [reply src/main/java/com/prowidesoftware/swift/constraints/BicConstraint.java:25]: > We normally do organize import with the IDE formatter and leave the order that is produced by it.

The import order is defined in Eclipse config file `org.eclipse.jdt.ui.prefs` (see above) This file treated `java|javax` imports differently. `jakarta` is like any other package name.

File: build.gradle (hunk 6 above)
Comment (sman-81) [reply build.gradle:39]: If the project even loosely follows [semantic versioning](https://semver.org/) rules, a breaking change requires a new major. It is not so much a question of benefit of a change but whether it is breaking.

Comment (zubri) [reply build.gradle:39]: Yes, major version would be changed from 9 to 10 to reflect the breaking change. 
The project does follow semantic versioning, with a custom prefix indicating the SRU.

Comment (essobedo) [reply build.gradle:39]: I think it is clear to everybody that a major version is needed for those changes, the question is more do you really want to support 2 new major versions (one for Java 11 + one for Jakarta as you proposed) instead of only one? Frameworks like Spring Boot, Quarkus, Camel, and many others, have one major version to do both (Java upgrade and Jakarta migration), as I did for this PR.

**Comment by essobedo** [comment]
OK I will propose 2 different PRs instead of one

**Comment by essobedo** [comment]
All remarks have been addressed

**Review by sman-81 — APPROVED** [review]
Looks good (requires Migration to Java 11 #152 to be merged first)

**Comment by essobedo** [comment]
@zubri Hi, Is it good enough for you? Is-there something missing?

**Comment by essobedo** [comment]
> > @zubri Hi, Is it good enough for you? Is-there something missing?
> 
> Yes. We are working on the code generation to produce the same out of the schemas. We have a gradle project using xjc and custom bindings. And also a custom Jaxb plugin. These have to be migrated as well to Jakarta and Java 11.

I see, thx for the information

**Comment by zubri** [comment]
> > > @zubri Hi, Is it good enough for you? Is-there something missing?
> > 
> > 
> > Yes. We are working on the code generation to produce the same out of the schemas. We have a gradle project using xjc and custom bindings. And also a custom Jaxb plugin. These have to be migrated as well to Jakarta and Java 11.
> 
> I see, thx for the information

If we face any issue we might contact you for help, but so far all good, we just need some time. We should be able to have the first alpha version in a couple of weeks at most.
